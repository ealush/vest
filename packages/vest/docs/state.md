# Understanding Vest's state

Vest is designed to help perform validations on user inputs. The nature of user inputs is that they are filled one by one by the user. In order to provide good user experience, the best approach is to validate fields as the user type, or when they leave the field.

The difficult part when validating upon user interaction is that we want to only validate the field that the user is currently interacting with, and not the rest of the form. This
This can be done with Vest's [`only()` hook](./exclusion). That's where the state mechanism is becoming useful.

When you have skipped fields in your validation suite, vest will try to see if those skipped fields ran in the previous suite, and merge them into the currently running suite result - so the result object you get will include all the fields that your user interacted with.

## What Vest's state does

- _Skipped field merge_

As mentioned before - whenever you skip a field, vest will look for it in your previously ran validations and add it to the current result.

- _Lagging async `done` callback blocking_

In case you have an async test that didn't finish from the previous suite run - and you already ran another async test for the same field - vest will block the [`done()`]('./result#done) callbacks for that field from running for the previous suite result.

## Resetting suite state with `.reset();`

In some cases, such as form reset, you want to discard of previous validation results. This can be done with `vest.reset()`.

`.reset` disables all pending async tests in your suite and empties the state out.

### Usage:

`.rese()` Is a property on your validation suite, calling it will remove your suite's state.

```js
import vest from 'vest';

const v = vest.create('suite_name', () => {
  // Your tests go here
});

v.reset(); // validation result is removed from Vest's state.
```
