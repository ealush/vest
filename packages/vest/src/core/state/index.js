import context from '../context';

const SYMBOL_ADD_TO_STATE = Symbol();

export default (function createState() {
  function registerHandler(initialValue) {
    let key;
    function use(patcher) {
      const { stateRef } = context.use();

      if (typeof patcher === 'function') {
        update(patcher);
      }

      function update(patcher) {
        const { stateRef } = context.use();
        const currentState = stateRef.current();
        const prevState = stateRef.prev();

        stateRef.set(
          key,
          typeof patcher === 'function'
            ? patcher(currentState?.[key], prevState?.[key])
            : patcher
        );
      }

      return [stateRef.current()[key], update];
    }

    use[SYMBOL_ADD_TO_STATE] = function (stateKey, ...args) {
      const { stateRef } = context.use();
      key = stateKey;

      if (!Object.prototype.hasOwnProperty.call(stateRef.current(), key)) {
        stateRef.set(
          key,
          typeof initialValue === 'function'
            ? initialValue(...args)
            : initialValue
        );
      }
    };

    return use;
  }

  function createRef(handlers = {}) {
    const state = [];

    function current() {
      return state[0];
    }

    function prev() {
      return state[1];
    }

    function set(key, value) {
      state[0] = {
        ...state[0],
        [key]: value,
      };
    }

    function reset() {
      state.length = 0;

      unshift();
    }

    function unshift() {
      context.run({ stateRef }, () => {
        state.unshift({});
        for (const key in handlers) {
          if (!Object.prototype.hasOwnProperty.call(current(), key)) {
            if (Array.isArray(handlers[key])) {
              /*
                state.createRef({
                  useSuiteId: [useSuiteId, [id, name]],
                })
              */
              handlers[key][0][SYMBOL_ADD_TO_STATE](key, ...handlers[key][1]);
            } else {
              /*
                state.createRef({
                  useSuiteId,
                });
              */
              handlers[key][SYMBOL_ADD_TO_STATE](key);
            }
          }
        }
      });
    }

    const stateRef = {
      current,
      prev,
      reset,
      set,
      unshift,
    };

    context.run({ stateRef }, reset);

    return stateRef;
  }

  return {
    registerHandler,
    createRef,
  };
})();
