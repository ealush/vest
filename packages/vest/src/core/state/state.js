import callEach from 'callEach';
import context from 'ctx';
import isFunction from 'isFunction';

export default (function createState() {
  function registerHandler(initialValue) {
    let key;
    function use(patcher) {
      const { stateRef, reg } = context.use();

      // Register state handler
      if (reg) {
        key = reg.key;
        if (!stateRef.current().hasOwnProperty(key)) {
          stateRef.set(
            key,
            isFunction(initialValue)
              ? initialValue.apply(null, reg.args)
              : initialValue
          );
        }
        return;
      }

      if (isFunction(patcher)) {
        update(patcher);
      }

      function update(patcher) {
        const { stateRef } = context.use();
        const currentState = stateRef.current();

        stateRef.set(
          key,
          isFunction(patcher) ? patcher(currentState[key]) : patcher
        );
      }

      return [stateRef.current()[key], update];
    }

    return use;
  }

  function createRef(handlers = {}) {
    let state = {};
    const registeredHandlers = [];

    function current() {
      return state;
    }

    function set(key, value) {
      Object.assign(state, { [key]: value });
    }

    function reset() {
      state.length = 0;

      state = {};
      callEach(registeredHandlers);
    }

    const stateRef = {
      current,
      reset,
      set,
    };

    for (const key in handlers) {
      const [handler, args] = [].concat(handlers[key]);
      registeredHandlers.push(
        context.bind({ stateRef, reg: { key, args } }, handler)
      );
    }

    context.run({ stateRef }, reset);

    return stateRef;
  }

  return {
    registerHandler,
    createRef,
  };
})();
