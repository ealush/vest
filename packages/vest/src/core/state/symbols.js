export const SYMBOL_STATE = Symbol('vestState');
export const SYMBOL_SUITES = Symbol('vestSuites');
export const SYMBOL_CANCELED = Symbol('canceledTests');
