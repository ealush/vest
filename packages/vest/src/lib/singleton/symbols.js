/**
 * @type {String} Vest's major version.
 */
const VEST_MAJOR = VEST_VERSION.split('.')[0];

/**
 * @type {Symbol} Used to store a global instance of Vest.
 */
export const SYMBOL_VEST = Symbol.for(`VEST#${VEST_MAJOR}`);
