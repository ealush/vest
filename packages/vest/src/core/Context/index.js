import {
  EXCLUSION_ITEM_TYPE_TESTS,
  EXCLUSION_ITEM_TYPE_GROUPS,
} from '../../hooks/exclusive/constants';
import singleton from '../../lib/singleton';

class Context {
  /**
   * Clears stored instance from constructor function.
   */
  static clear() {
    const ctx = singleton.useContext();

    if (ctx?.parentContext) {
      singleton.use().ctx = ctx.parentContext;
      ctx.parentContext.removeChildContext();
    } else {
      singleton.use().ctx = null;
    }
  }

  /**
   * Creates a new context object, and assigns it as a static property on Vest's singleton.
   * @param {Object} ctxRef   Context data reference.
   * @returns {Context} either an existing or a new context object.
   */
  constructor(ctxRef) {
    const ctx = singleton.useContext();
    Object.assign(this, ctxRef);

    if (ctx) {
      ctx.setChildContext(this);
    }

    singleton.use().ctx = this;
  }

  get suiteId() {
    return this.lookup('suite_id');
  }

  set suiteId(suiteId) {
    this.suite_id = suiteId;
  }

  get currentTest() {
    return this.lookup('current_test');
  }

  set currentTest(testObject) {
    this.current_test = testObject;
  }

  get groupName() {
    return this.lookup('group_name');
  }

  set groupName(groupName) {
    this.group_name = groupName;
  }

  get exclusion() {
    let key = this.lookup('_exclusion');

    if (key === undefined) {
      key = this._exclusion = {
        [EXCLUSION_ITEM_TYPE_TESTS]: {},
        [EXCLUSION_ITEM_TYPE_GROUPS]: {},
      };
    }

    return key;
  }

  lookup(key) {
    let ctx = this;
    do {
      if (ctx[key]) {
        return ctx[key];
      }
      ctx = ctx.parentContext;
    } while (ctx);
  }

  setParentContext(parentContext) {
    this.parentContext = parentContext;
  }

  setChildContext(childContext) {
    childContext.parentContext = this;
    this.childContext = childContext;
  }

  removeChildContext() {
    this.childContext = null;
  }
}

export default Context;
