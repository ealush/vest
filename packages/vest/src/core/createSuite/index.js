import { OPERATION_MODE_STATEFUL } from '../../constants';
import runWithContext from '../../lib/runWithContext';
import singleton from '../../lib/singleton';
import throwError from '../../lib/throwError';
import produce from '../produce';
import getSuiteState from '../state/getSuiteState';
import registerSuite from '../state/registerSuite';
import { mergeSkipped } from '../test/lib/skipped';
import runAsyncTest from '../test/runAsyncTest';
import { SUITE_INIT_ERROR } from './constants';

/**
 * Initializes a validation suite, creates a validation context.
 * @param {String} name     Identifier for validation suite.
 * @param {Function} tests  Validation suite body.
 * @returns {Function} validator function.
 */
const createSuite = (name, tests) => {
  if (typeof name !== 'string') {
    return throwError(
      SUITE_INIT_ERROR + ' Expected name to be a string.',
      TypeError
    );
  }

  if (typeof tests !== 'function') {
    return throwError(
      SUITE_INIT_ERROR + ' Expected tests to be a function.',
      TypeError
    );
  }

  // returns validator function
  return (...args) => {
    const parentContext = singleton.useContext() ?? {
      name,
      tests,
      suiteId: name,
      operationMode: OPERATION_MODE_STATEFUL,
    };

    const output = runWithContext(parentContext, context => {
      registerSuite();
      const { suiteId } = context;
      tests.apply(null, args);
      mergeSkipped(suiteId);

      [...getSuiteState(suiteId).pending].forEach(runAsyncTest);
      return produce(getSuiteState(suiteId));
    });

    return output;
  };
};

export default createSuite;
