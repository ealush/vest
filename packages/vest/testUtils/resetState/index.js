import { OPERATION_MODE_STATEFUL } from '../../src/constants';
import * as state from '../../src/core/state';
import getState from '../../src/core/suite/getState';
import runRegisterSuite from '../runRegisterSuite';

const resetState = (suiteId, suiteName = suiteId) => {
  state.set(() => null);
  state.register();

  if (suiteName) {
    runRegisterSuite({
      suiteId,
      name: suiteName,
      operationMode: OPERATION_MODE_STATEFUL,
    });
    return getState(suiteId);
  }
};

export default resetState;
