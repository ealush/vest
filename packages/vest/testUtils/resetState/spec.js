import { getState } from '../../src/core/state';
import { SYMBOL_SUITES } from '../../src/core/state/symbols';
import resetState from '.';

describe('resetState', () => {
  test('Sanity', () => {
    expect(getState()).toMatchSnapshot();
  });

  it('Should create a new state', () => {
    const initialState = getState();
    expect(getState()).toBe(initialState);
    resetState();

    expect(getState()).not.toBe(initialState);
  });

  describe('When invoked with suite name', () => {
    let suiteId;

    beforeEach(() => {
      suiteId = 'test-suite';
    });

    it('Should add a new suite to the state', () => {
      expect(getState(SYMBOL_SUITES)).toEqual({});
      resetState(suiteId);
      expect(getState(SYMBOL_SUITES)[suiteId]).toBeTruthy();
      expect(getState(SYMBOL_SUITES)[suiteId]).toMatchSnapshot();
    });
  });
});
