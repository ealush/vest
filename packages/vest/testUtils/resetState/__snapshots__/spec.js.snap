// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resetState Sanity 1`] = `
Object {
  "_canceled": Object {},
  "_suites": Object {},
}
`;

exports[`resetState When invoked with suite name Should add a new suite to the state 1`] = `
Array [
  Object {
    "doneCallbacks": Array [],
    "fieldCallbacks": Object {},
    "groups": Object {},
    "lagging": Array [],
    "name": "test-suite",
    "pending": Array [],
    "suiteId": "test-suite",
    "testObjects": Array [],
    "tests": Object {},
  },
  undefined,
]
`;
